File: core/treesitter/universal_ast_normalizer.py
from __future__ import annotations

from typing import Any, Dict, List


MAX_AST_NODES = 10000


def normalize_ast(
    raw_nodes: List[Dict[str, Any]],
    language: str,
) -> Dict[str, Any]:

    nodes = []

    edges = []

    for i, node in enumerate(
        raw_nodes[:MAX_AST_NODES]
    ):

        node_id = f"{language}_{i}"

        nodes.append({
            "id": node_id,
            "type": node.get("type"),
            "language": language,
        })

        parent = node.get("parent")

        if parent is not None:

            edges.append({
                "from": parent,
                "to": node_id,
                "relation": "ast_edge",
            })

    return {
        "nodes": nodes,
        "edges": edges,
        "language": language,
        "bounded": True,
    }

File: src/treesitter/universalAstNormalizer.ts
/**
 * Converted from Python: core/treesitter/universal_ast_normalizer.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let MAX_AST_NODES: any = 10000;
export function normalizeAst(raw_nodes: any, language: any): any {
  var nodes: any[] = [];
  var edges: any[] = [];
  var i: any;
  var node: any;
  for ([i, node] of py.enumerate(py.slice(raw_nodes, null, MAX_AST_NODES))) {
    var node_id: any = `${py.toStr(language)}_${py.toStr(i)}`;
    py.listAppend(nodes, {"id": node_id, "type": py.get(node, "type"), "language": language});
    var parent: any = py.get(node, "parent");
    if ((parent !== null && parent !== undefined)) {
      py.listAppend(edges, {"from": parent, "to": node_id, "relation": "ast_edge"});
    }
  }
  return {"nodes": nodes, "edges": edges, "language": language, "bounded": true};
}
